[ACTV-10229]Invoice Post Processing update for DU Template

swittenberger opened the pull request:
User Story:
https://uipath.atlassian.net/browse/ACTV-10229

This feature will add the ability to post process Invoices to reduce the number of hits to Action Center

Commit:
Invoice Post Processing update for DU Template

Commit:
Updated annotations.

alexandru-luca reviewed (changes requested):
Clear all breakpoints
Rename AlwaysValidateExtractionWithActionCenter setting/asset name to AlwaysValidateExtraction or something similar - it's the same Setting/Asset for both Attended & Unattended, "ActionCenter" in its name will cause confusion
55_ExtractionBussinessRuleValidation.xaml:
 - add an "Execution Start" log message
 - usage of the ClassificationResult argument is not ideal; for consistency, io_ExtractionResults.ResultsDocument.DocumentTypeId inside should be used instead (if we ever get an extractor smart enough to change the classification type)

InvoicePostProcessing.xaml:
 - rename it to InvoicePostProcessingExample - to make it clear that this is an example to be customised as needed
  - add a workflow annotation explaining the Business Rules that have been implemented AND the overall validation logic;  put a HUGE disclamer that this is our GENERIC example and should be customised as needed, not blindly used as-is 
- document the code - few people will read the manual, we need to make it easy for devs to understand what's going on just by looking at the code
 - add some annotations giving details on every important step: "Setting Up Variables", "Check Mandatory Extracted Values", "Value Cleanup", "Line Verification", "Extraction Results Check"
 - rename "Comment" activity to something more descriptive
 - to improve readability, rename "For Each Item in the Taxonomy" to "For Each Field in the Taxonomy" and rename its local variable from "item" to "field"
 - rename "Line Verification" to "For Each Row - Line Verification" to improve readability
 - "Line Verification" -> ""Compute Line Amount" -> Else branch - does it make sense to continue if a discrepancy is found? I think a "Break" activity should be added here
 - for the Math.Round(x, decimals0: the decimals value should NOT be hardcoded - make it a setting instead
 - OCR errors must be accounted for whenever parsing extracted text into a number - when using cdec()
 - Subtotal calculations - I'm a bit worried about rounding at every step; the subtotal should be calculated first and only rounded once, at the end
 

55_ExtractionBussinessRuleValidation.xaml & InvoicePostProcessing.xaml - please reorder Arguments to have them in a neat, logical order: in (in_Config should be last), io, out

Future improvement points:
- read the "InvoicePostProcessing" settings only when needed, not every time
- update the Classification logic to follow the same pattern as the data extraction - "AlwaysValidateClassification" setting and a dedicated workflow for checking classification business rules

dragospa committed:
Updated the rest of the annotations.

Commit:
- Cleared all breakpoints
- added an "Execution Start" log message for both Classification and Extraction Business Rule Validations
- fixed usage of the ClassificationResult, using io_ExtractionResults.ResultsDocument.DocumentTypeId instead
- Renamed AlwaysValidateExtractionWithActionCenter setting/asset to AlwaysValidateExtraction
- Added AlwaysValidateClassification
- added a workflow annotation explaining the Business Rules that have been implemented AND the overall validation logic;
- added a HUGE disclaimer and documentation on how to use the Invoice Post Processing
- documented the code
- added some annotations giving details on every important step: "Setting Up Variables", "Check Mandatory Extracted Values", "Value Cleanup", "Line Verification", "Extraction Results Check"
- renamed "Comment" activity to Comment Invoice Post Processing Logic
- renamed "For Each Item in the Taxonomy" to "For Each Field in the Taxonomy"
- renamed local variable from "item" to "field"
- renamed "Line Verification" to "For Each Row - Line Verification" to improve readability
- added a break activity for the row computation
- added Math.Round Decimal in the config file for easier changing
- removed all roundings, except in the beginning when computing Quantity * Unit Price
- reordered Arguments to have them in a neat, logical order: in (in_Config should be last), io, out
- updated the Classification logic to follow the same pattern as the data extraction -
- added "AlwaysValidateClassification" setting and a dedicated workflow for checking classification business rules

swittenberger commented:
Updated, and tested, based on comments.

Two things here:
- I do not agree with changing the name to InvoicePostProcessingExample. While they will be using this as a template for their specific development, having the Example word there doesn't help with anything, only that they would need to take an extra step to remove that when moving to production/dev
- OCR errors should be accounted for - While I agree in principle, if there are errors here, the workflow should fail, and not try to correctly guess what the OCR wanted to say. Right now, if there are weird characters (at the line items), they will be removed. If the extracted values do not match, then the process will fail as intended.

Commit:
- fixed small bugs

Commit:
- fixed another small bug

Commit:
- fixed another small bug

Commit:
- Added two s-es

alexandru-luca committed:
Small cosmetic improvements
